### PR TITLE
Read from Google docs and write content to shared directory

### PR DESCRIPTION
**Pull request by matthewsilver**
TODO's for future iterations:

- Abstract hard-coded values and environmental variables to config(s)
- Pull down/push to Github only documents that have been modified. Might be reasonable with https://developers.google.com/drive/api/v3/change-overview.
     > Is there something asynchronous? Like any Google doc change automatically triggers an event.
- Clean up ingestion code for better named variables, more modular, refactored etc.
- Add documentation explaining Google API setup, Git SSH key, running the ingestion, how the code works, expected input and output, etc.
- If no docs have changed, don't automatically commit code
- Change directory structure to accommodate for Serge (see Aditya's comments)

### CODE REVIEW

File: shared_directory/Raw_Document_1.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"title": "Raw_Document_1", "content": "We use technology to make New York ready for the future.\n\nWe can adapt to change and unexpected situations.\n\nWe are diverse and agile.\n\nThis is a new change.\n\nThis is another change.\n"}
```
Comment (as1729) [inline shared_directory/Raw_Document_1.json]: This is great!

For the "shared" folder directory structure, I'm thinking we can organize it similar to how smartcat does it-

```
/shared_directory
    /en
        raw_document_1.json
    /ru
        raw_document_1.json.ru
```

All of the `en/` files gets populated directly from google drive. And I can setup `serge` to output inside of the each language prefix. Serge _should_ generate those folders for us automatically.

**Review by sladymon — COMMENTED** [review]
Question: In read_structural_elements() it looks like all the content from various elements is being grouped together as a single text. How will the localized versions be rebuilt with the same element structure?

File: shared_directory/Raw_Document_1.json (hunk 1 above)
Comment (matthewsilver) [reply shared_directory/Raw_Document_1.json]: @as1729 sounds like a good idea!

**Comment by matthewsilver** [comment]
@sladymon that's a good question -- with the JSON parser, I think Serge could handle reading in more complex structures, but in terms of how the Serge translation writes that information back that's something we might need to look into.

This also brings up a broader point of QA and how to deal with preserving formatting when rendering content (and what to do if some content is unrenderable for whatever reason)

File: ingestion/ingest-google-docs.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        Args:
+            element: a ParagraphElement from a Google Doc.
+    """
+    text_run = element.get('textRun')
+    if not text_run:
+        return ''
+    return text_run.get('content')
+
+
+def read_strucutural_elements(elements):
```
Comment (liquidsteves) [inline ingestion/ingest-google-docs.py]: Spelling error on "structural"

File: ingestion/ingest-google-docs.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    # Enumerate over Google doc ids and fetch content
+    store = file.Storage('secrets/token.json')
+    creds_doc = store.get()
+    if not creds_doc or creds_doc.invalid:
+        flow_doc = client.flow_from_clientsecrets('secrets/credentials.json', SCOPES_DOC)
+        creds_doc = tools.run_flow(flow_doc, store)
+    service_docs = discovery.build('docs', 'v1', http=creds_doc.authorize(Http()), discoveryServiceUrl=DISCOVERY_DOC)
+    for file_id in file_ids:
+        result = service_docs.documents().get(documentId=file_id).execute()
```
Comment (liquidsteves) [inline ingestion/ingest-google-docs.py:136]: "documentId=" is just a readability thing to explain what file_id is doing?

**Review by liquidsteves — APPROVED** [review]
I vote that it's good to merge this in as soon as Matt feels it's ready. Lots more to be done, but better to keep merges small. How does everyone else feel?

File: ingestion/ingest-google-docs.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                    text += read_strucutural_elements(cell.get('content'))
+        elif 'tableOfContents' in value:
+            # The text in the TOC is also in a Structural Element.
+            toc = value.get('tableOfContents')
+            text += read_strucutural_elements(toc.get('content'))
+    return text
+
+
+def run():
+    # Find ids of all Google docs in the raw Serge folder
```
Comment (liquidsteves) [inline ingestion/ingest-google-docs.py]: Move this comment down to the while True loop, and instead comment here that you are authenticating.

File: ingestion/ingest-google-docs.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        # If there are no (valid) credentials available, let the user log in.
+    if not creds or creds.invalid:
+        if creds and creds.expired and creds.refresh_token:
+            creds.refresh(Request())
+        else:
+            flow = client.flow_from_clientsecrets('secrets/credentials.json', SCOPES)
+            creds = tools.run_flow(flow, store)
+        # Save the credentials for the next run
+        with open('secrets/token_read.pickle', 'wb') as token:
+            pickle.dump(creds, token)
```
Comment (liquidsteves) [inline ingestion/ingest-google-docs.py]: Why do we pickle these credentials for the next run, but not creds_doc below?

Btw, can we move this credential logic to a helper method?

File: ingestion/ingest-google-docs.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            # The text in the TOC is also in a Structural Element.
+            toc = value.get('tableOfContents')
+            text += read_strucutural_elements(toc.get('content'))
+    return text
+
+
+def run():
+    # Find ids of all Google docs in the raw Serge folder
+    store = file.Storage('secrets/token_read.json')
+    page_token = None
```
Comment (liquidsteves) [inline ingestion/ingest-google-docs.py]: Don't declare this until the start of the while True loop

Also, having two different meanings of "token" here is confusing. Can we rename page_token to something else? Or add a comment "The page token is totally unrelated to the oath token"

File: ingestion/ingest-google-docs.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            # Process change
+            file_id = file_content.get('id')
+            print('Found file: %s (%s)' % (file_content.get('name'), file_id))
+            file_ids.append(file_id)
+        page_token = response.get('nextPageToken', None)
+        if page_token is None:
+            break
+
+    # Enumerate over Google doc ids and fetch content
+    store = file.Storage('secrets/token.json')
```
Comment (liquidsteves) [inline ingestion/ingest-google-docs.py]: Declare new variable store_doc.

Could token.json be named token_write.json?

File: ingestion/ingest-google-docs.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+            text += read_strucutural_elements(toc.get('content'))
+    return text
+
+
+def run():
+    # Find ids of all Google docs in the raw Serge folder
+    store = file.Storage('secrets/token_read.json')
+    page_token = None   
+    creds = None
+    flow = client.flow_from_clientsecrets('secrets/credentials.json', SCOPES)
```
Comment (liquidsteves) [inline ingestion/ingest-google-docs.py]: Delete this line

File: ingestion/ingest-google-docs.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            table = value.get('table')
+            for row in table.get('tableRows'):
+                cells = row.get('tableCells')
+                for cell in cells:
+                    text += read_strucutural_elements(cell.get('content'))
+        elif 'tableOfContents' in value:
+            # The text in the TOC is also in a Structural Element.
+            toc = value.get('tableOfContents')
+            text += read_strucutural_elements(toc.get('content'))
+    return text
```
Comment (liquidsteves) [inline ingestion/ingest-google-docs.py]: Can you write a unit test for read_structural_elements ?

**Review by liquidsteves — COMMENTED** [review]
Sorry for the comments trickling in at strange times. My python is really rusty so I figured it would be good for me to read over the code a second time as my brain figures out how to process it.

File: ingestion/ingest-google-docs.py (hunk 3 above)
Comment (matthewsilver) [reply ingestion/ingest-google-docs.py:136]: yep

File: ingestion/ingest-google-docs.py (hunk 5 above)
Comment (matthewsilver) [reply ingestion/ingest-google-docs.py]: I refactored this all to use the same authentication flow for both docs and drive and made it a method outside of `run()`

File: ingestion/ingest-google-docs.py (hunk 7 above)
Comment (matthewsilver) [reply ingestion/ingest-google-docs.py]: Yep, cleared up the naming for all this

File: ingestion/ingest-google-docs.py (hunk 6 above)
Comment (matthewsilver) [reply ingestion/ingest-google-docs.py]: Yeah, I renamed this stuff and refactored into a `generate_secrets()` method

File: ingestion/ingest-google-docs.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        'secrets/token_read_drive.json',
+        'secrets/credentials.json',
+        SCOPE_READ_DRIVE
+        )
+    service_docs = generate_secrets(
+        'secrets/token_read_docs.pickle',
+        'secrets/token_read_docs.json',
+        'secrets/credentials.json',
+        SCOPE_READ_DOCS
+        )
```
Comment (liquidsteves) [inline ingestion/ingest-google-docs.py:113]: Thanks! This is a lot cleaner.

File: ingestion/ingest-google-docs.py (hunk 9 above)
Comment (matthewsilver) [reply ingestion/ingest-google-docs.py]: Made a skeleton for it, but the actual format of the JSON that will be written to the shared directory will probably change soon (as discussed, preserve the formatting of the doc) so I'll hold off for now

File: ingestion/ingest-google-docs.py (hunk 10 above)
Comment (matthewsilver) [reply ingestion/ingest-google-docs.py:113]: For a future PR I'll put this and other constants in a YAML or something

File: ingestion/ingest-google-docs.py (hunk 9 above)
Comment (liquidsteves) [reply ingestion/ingest-google-docs.py]: Sounds good.

**Comment by liquidsteves** [comment]
I thought the master branch was just where we kept README files. Lol

Woot! Code is merging!

**Pull request merged**